src/Rules/AbstractNodeVisitorRule.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+declare(strict_types=1);
+
+namespace TwigCsFixer\Rules;
+
+use Twig\Environment;
+use Twig\Node\Node;
+use TwigCsFixer\Report\Report;
+use TwigCsFixer\Report\Violation;
+use TwigCsFixer\Report\ViolationId;
+use Webmozart\Assert\Assert;
+
+abstract class AbstractNodeVisitorRule implements NodeVisitorRuleInterface
+{
+    private ?Report $report = null;
+
+    private ?string $filePath = null;
+
+    /**
+     * @var list<ViolationId>
+     */
+    private array $ignoredViolations = [];
+
+    public function getName(): string
+    {
+        return static::class;
+    }
+
+    public function getShortName(): string
+    {
+        $shortName = (new \ReflectionClass($this))->getShortName();
+
+        return str_ends_with($shortName, 'Rule') ? substr($shortName, 0, -4) : $shortName;
+    }
+
+    public function lintFile(Report $report, string $filePath, array $ignoredViolations = []): void
+    {
+        $this->report = $report;
+        $this->filePath = $filePath;
+        $this->ignoredViolations = $ignoredViolations;
+    }
+
+    public function enterNode(Node $node, Environment $env): Node
+    {
+        return $node;
+    }
+
+    public function leaveNode(Node $node, Environment $env): Node
+    {
+        return $node;
+    }
+
+    public function getPriority(): int
+    {
+        return 0;
+    }
+
+    protected function addWarning(string $message, ?int $line = null, ?string $messageId = null): bool
+    {
+        Assert::notNull($this->filePath);
+
+        return $this->addMessage(
+            Violation::LEVEL_WARNING,
+            $message,
+            $this->filePath,
+            $line,
+            null,
+            $messageId,
+        );
+    }
+
+    protected function addFileWarning(string $message, ?int $line = null, ?string $messageId = null): bool
+    {
+        Assert::notNull($this->filePath);
+
+        return $this->addMessage(
+            Violation::LEVEL_WARNING,
+            $message,
+            $this->filePath,
+            null,
+            null,
+            $messageId,
+        );
+    }
+
+    protected function addError(string $message, ?int $line = null, ?string $messageId = null): bool
```

Comment:
I wonder if we should provide the signature
```
protected function addError(string $message, Node $node, ?string $messageId = null): bool
```
and when we're using
```
$node->getTemplateLine();
$node->getTemplateName();
```

Reply:
$node->getTemplateName() is nullable.